src/pudl/output/pudltabl.py

```diff
@@ -1207,6 +1207,62 @@ def epacamd_eia(
             self._dfs["epacamd_eia"] = pudl.output.epacems.epacamd_eia(self.pudl_engine)
         return self._dfs["epacamd_eia"]
 
+    ###########################################################################
+    # FOR PICKLING AND OTHER IO
+    ###########################################################################
+
+    def __getstate__(self) -> dict:
+        """Get current object state for pickle.
+
+        This method is run as part of pickling the object. It needs to return the
+        object's current state with any un-serializable objects converted to a form that
+        can be serialized. See :meth:`object.__getstate__` for further details on the
+        expected behavior of this method.
+        """
+        return self.__dict__.copy() | {
+            # defaultdict may be serializable but lambdas are not, so it must go
+            "_dfs": dict(self.__dict__["_dfs"]),
+            # sqlalchemy engines are also a problem here, saving the URL should
+            # provide enough of what is needed to recreate it, though that means the
+            # pickle is not portable, but any fix to that should probably happen
+            # on the other side
+            "pudl_engine": str(self.__dict__["pudl_engine"].url)
+            .removeprefix("Engine(")
+            .removesuffix(")"),
+        }
+
+    def __setstate__(self, state: dict) -> None:
+        """Restore the object's state from a dictionary.
+
+        This method is run when the object is restored from a pickle. Anything
+        that was changed in :meth:`pudl.output.pudltabl.PudlTabl.__getstate__` must be
+        undone here. Another important detail is that ``__init__`` is not run when an
+        object is de-serialized, so under some circumstances, setup that happened there
```

Comment:
What circumstances are these?

Reply:
I think usually circumstances where `__init__` does things outside the class like changing globals, opening a file, etc. I altered the text to be less mysterious.

 I put this comment for two reasons (1) for anyone looking at this method that wasn't familiar with how pickle restores objects because it's not immediately obvious how this fits into the rest that process, and (2) in case something special needs to be done for `Datastore`. Granted (2) is sort of redundant given that is better and more clearly addressed by the PR narrative. I think I was just a little paranoid about breaking `Datastore` resulting in odd behavior.